### ISSUE TITLE
Missing buildLocalizeArrayFn module

### ISSUE THREAD

**Issue by ericreis** [date-fns/date-fns#1014]
I was trying to dynamically import locales from `date-fns` package and bumped into the following error: 

`Module not found: Can't resolve '../../../_lib/buildLocalizeArrayFn/index.js' in '/Users/ericreis/dev/vtex/test/node_modules/date-fns/locale/ar-DZ/_lib/localize'`

Here is the code I'm using to perform the dynamic import:

```js
const locale = 'en-GB'
import(`date-fns/locale/${locale}/index.js`).then(
  data => {
    console.log(data)
  })
```

I searched everywhere for this `buildLocalizeArrayFn`folder, but I didn't find it anywhere. Is it a bug? If not, could anyone point to me what I am doing wrong?

**Comment by TwoDCube** [comment]
Because you used `en-GB` I assume you're using `date-fns@next`. By doing your import like this, it tries to find every file named `index.js` under the `locale` directory (even those with depth > 1!, the locale variable can contain a `/`). And there are some languages containing old/unused code which are not imported by date-fns but your dynamic import tries to use them too thus giving you errors.

**Comment by ericreis** [comment]
I am using v2.0.0alpha23. What would be the correct way to dinamically import the package then?

**Comment by ericreis** [comment]
But my point actually is: Does this `'../../../_lib/buildLocalizeArrayFn/index.js'` require (for example in line 2 [here](https://github.com/date-fns/date-fns/blob/master/src/locale/ar-DZ/_lib/localize/index.js)) make any sense? I think it does not since there is no sign of this path (`buildLocalizeArrayFn` folder) anywhere in the project (even in old versions).

**Comment by TwoDCube** [comment]
I currently couldn't find a way to import them properly. At the weekend I'll create a PR to export the languages from single files instead of folders so this depth thing won't be a problem. (Something like angular does [here](https://github.com/angular/angular/tree/master/packages/common/locales))

So you would write something like:
```javascript
import(`date-fns/locales/${locale}.js`)
```

This way there are no folders and depths to worry about. What do you think?

**Comment by TwoDCube** [comment]
There was a folder like that, see [this](https://github.com/date-fns/date-fns/tree/0ff256de6b260e50daf40be15581d07c7afcdaed/src/locale/_lib/buildLocalizeArrayFn)

**Comment by ericreis** [comment]
> This way there are no folders and depths to worry about. What do you think?

I personally prefer this way too.


> There was a folder like that, see [this](https://github.com/date-fns/date-fns/tree/0ff256de6b260e50daf40be15581d07c7afcdaed/src/locale/_lib/buildLocalizeArrayFn)

Maybe that's the folder that is missing now.

**Comment by ericreis** [comment]
I'll try adding that folder to my current test project and post the updates.

**Comment by ericreis** [comment]
I think I managed to fix all the errors by adding the `buildLocalizeArrayFn`, `buildParseFn`, `parseDecimal`and `build_formatting_tokens_reg_exp` folders. This last one I assumed that it used to be [this](https://github.com/date-fns/date-fns/tree/0ff256de6b260e50daf40be15581d07c7afcdaed/src/locale/_lib/buildTokensRegExp).

**Comment by TwoDCube** [comment]
I don't think it's a great idea to just recover those deleted directories. Probably they had a reason to remove those folders.

Also I think it would still try to include the index.js -s from deeper directories too. I'll also make a PR to show my idea too.

**Comment by ericreis** [comment]
> I don't think it's a great idea to just recover those deleted directories. Probably they had a reason to remove those folders.

I totally agree with you. I made those changes just to check if we would have any other errors regarding dynamic imports and it appears we don't.

> Also I think it would still try to include the index.js -s from deeper directories too.

 Yeah, it does.

> I'll also make a PR to show my idea too.

I personally think that the idea of having all `${locale}.js` files in a single folder with no deeper levels is the way to go. If you need any help in the PR, just let me know.

**Comment by kossnocorp** [comment]
> I personally think that the idea of having all ${locale}.js files in a single folder with no deeper levels is the way to go. If you need any help in the PR, just let me know.

@ericreis please don't do that. I don't have time right now to dive deep into your problem, but this solution won't be accepted.

First of all, there are unsupported locales that you don't want to import anyway. You can see the list here: https://github.com/date-fns/date-fns/blob/master/outdatedLocales.json. Secondly, the locales should stay modular.

**Comment by ericreis** [comment]
@kossnocorp perhaps then these locales should me moved to a separate location? Dynamic import is a powerful feature that right now is breaking due to outdated files in modules that won’t even be called.

**Comment by developdeez** [comment]
Any solution to this? I have the same issue.

**Comment by san-pblr-gct** [comment]
I am getting compilation error due to unsupported locales present inside locale folder. Any solution?

**Comment by san-pblr-gct** [comment]
I did a work around.

I created supportingLanguages.js with

const supportedLanguages = {
  af: async () => await import("date-fns/locale/af/index.js"),
  arSA: async () => await import("date-fns/locale/ar-SA/index.js"),
  bn: async () => await import("date-fns/locale/bn/index.js"),
  de: async () => await import("date-fns/locale/de/index.js"),
  el: async () => await import("date-fns/locale/el/index.js"),
  enCA: async () => await import("date-fns/locale/en-CA/index.js"),
  enGB: async () => await import("date-fns/locale/en-GB/index.js"),
  enUS: async () => await import("date-fns/locale/en-US/index.js"),
  eo: async () => await import("date-fns/locale/eo/index.js"),
  es: async () => await import("date-fns/locale/es/index.js"),
  et: async () => await import("date-fns/locale/et/index.js"),
  fr: async () => await import("date-fns/locale/fr/index.js"),
  gl: async () => await import("date-fns/locale/gl/index.js"),
  he: async () => await import("date-fns/locale/he/index.js"),
  hu: async () => await import("date-fns/locale/hu/index.js"),
  it: async () => await import("date-fns/locale/it/index.js"),
  ja: async () => await import("date-fns/locale/ja/index.js"),
  lt: async () => await import("date-fns/locale/lt/index.js"),
  nb: async () => await import("date-fns/locale/nb/index.js"),
  nl: async () => await import("date-fns/locale/nl/index.js"),
  pt: async () => await import("date-fns/locale/pt/index.js"),
  ptBR: async () => await import("date-fns/locale/pt-BR/index.js"),
  ru: async () => await import("date-fns/locale/ru/index.js"),
  sv: async () => await import("date-fns/locale/sv/index.js"),
  uk: async () => await import("date-fns/locale/uk/index.js"),
  vi: async () => await import("date-fns/locale/vi/index.js"),
  zhCN: async () => await import("date-fns/locale/zh-CN/index.js")
};

export default supportedLanguages;

and then where ever i wanted to import

 supportedLanguages[this.props.lang]().then(localeFile => {
     //do something
    });

**Comment by aliciacatalina** [comment]
I'm facing the same issue when trying to dynamically import a locale file, are there any plans to remove the outdated locales?

**Comment by kossnocorp** [comment]
I've released a fix for this issue as `v2.0.0-alpha.34`: https://gist.github.com/kossnocorp/a307a464760b405bb78ef5020a4ab136#v200-alpha34
